From aac1bce254425719438d36c9b0e6d4e5f506c01b Mon Sep 17 00:00:00 2001
From: Bones <babblebones@protnmail.com>
Date: Tue, 12 Aug 2025 14:55:32 -0400
Subject: [PATCH] [HACK] kernelbase: Replace stderr of yt-dlp process.

---
 dlls/kernelbase/process.c | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/dlls/kernelbase/process.c b/dlls/kernelbase/process.c
index 93554095a48..362edc8afc3 100644
--- a/dlls/kernelbase/process.c
+++ b/dlls/kernelbase/process.c
@@ -553,6 +553,25 @@ static BOOL product_name_matches(const WCHAR *app_name, const char *match)
     return TRUE;
 }
 
+static BOOL is_ytdlp( const WCHAR *app_name )
+{
+    static const WCHAR ytdlp_exeW[] = L"yt-dlp.exe";
+    static const size_t ytdlp_exeW_len = ARRAY_SIZE(ytdlp_exeW) - 1;
+    size_t len;
+
+    if (!app_name)
+        return FALSE;
+
+    len = wcslen( app_name );
+    if (len < ytdlp_exeW_len + 1)
+        return FALSE;
+    if (app_name[len - ytdlp_exeW_len - 1] != '\\' && app_name[len - ytdlp_exeW_len - 1] != '/')
+        return FALSE;
+    if (wcscmp( app_name + (len - ytdlp_exeW_len), ytdlp_exeW ))
+        return FALSE;
+    return TRUE;
+}
+
 static int battleye_launcher_redirect_hack( const WCHAR *app_name, WCHAR *new_name, DWORD new_name_len,
                                             WCHAR **orig_app_name )
 {
@@ -719,6 +738,9 @@ BOOL WINAPI DECLSPEC_HOTPATCH CreateProcessInternalW( HANDLE token, const WCHAR
     if (battleye_launcher_redirect_hack( app_name, name, ARRAY_SIZE(name), &orig_app_name ))
         app_name = name;
 
+    if (is_ytdlp( app_name ))
+        startup_info->hStdError = INVALID_HANDLE_VALUE;
+
     /* Warn if unsupported features are used */
 
     if (flags & (IDLE_PRIORITY_CLASS | HIGH_PRIORITY_CLASS | REALTIME_PRIORITY_CLASS |
-- 
2.50.1

